node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,839 @@
+					"Processing approvals from a peer",
+				);
+				for approval_vote in approvals.into_iter() {
+					self.import_and_circulate_approval(
+						ctx,
+						metrics,
+						MessageSource::Peer(peer_id.clone()),
+						approval_vote,
+					).await;
+				}
+			}
+		}
+	}
+
+	async fn handle_peer_view_change(
+		&mut self,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		metrics: &Metrics,
+		peer_id: PeerId,
+		view: View,
+	) {
+		Self::unify_with_peer(&mut self.blocks, ctx, metrics, peer_id.clone(), view.clone()).await;
+		let finalized_number = view.finalized_number;
+		self.peer_views.insert(peer_id.clone(), view);
+
+		// cleanup
+		let blocks = &mut self.blocks;
+		self.blocks_by_number
+			.range(0..=finalized_number)
+			.map(|(_n, h)| h)
+			.flatten()
+			.for_each(|h| {
+				if let Some(entry) = blocks.get_mut(h) {
+					entry.known_by.remove(&peer_id);
+				}
+			});
+	}
+
+	async fn handle_our_view_change(
+		&mut self,
+		_metrics: &Metrics,
+		view: View,
+	) {
+		// split_off returns everything after the given key, including the key
+		let split_point = view.finalized_number.saturating_add(1);
+		let mut old_blocks = self.blocks_by_number.split_off(&split_point);
+		std::mem::swap(&mut self.blocks_by_number, &mut old_blocks);
+
+		old_blocks.values()
+			.flatten()
+			.for_each(|h| {
+				self.blocks.remove(h);
+			});
+	}
+
+	async fn import_and_circulate_assignment(
+		&mut self,
+		ctx: &mut impl SubsystemContext<Message = ApprovalDistributionMessage>,
+		_metrics: &Metrics,
+		source: MessageSource,
+		assignment: IndirectAssignmentCert,
+		claimed_candidate_index: CandidateIndex,
+	) {
+		let block_hash = assignment.block_hash.clone();
+		let validator_index = assignment.validator;
+
+		let entry = match self.blocks.get_mut(&block_hash) {
+			Some(entry) => entry,
+			None => {
+				if let Some(peer_id) = source.peer_id() {
+					modify_reputation(ctx, peer_id, COST_UNEXPECTED_MESSAGE).await;
+				}
+				return;
+			}
+		};
+
+		// compute a fingerprint of the assignment
+		let fingerprint = MessageFingerprint::Assignment(
+			block_hash,
+			claimed_candidate_index,
+			validator_index,
+		);
+
+		if let Some(peer_id) = source.peer_id() {
+			// check if our knowledge of the peer already contains this assignment
+			match entry.known_by.entry(peer_id.clone()) {
+				hash_map::Entry::Occupied(knowledge) => {
+					if knowledge.get().known_messages.contains(&fingerprint) {
+						modify_reputation(ctx, peer_id, COST_DUPLICATE_MESSAGE).await;
+						return;
+					}
+				}
+				hash_map::Entry::Vacant(_) => {
+					modify_reputation(ctx, peer_id.clone(), COST_UNEXPECTED_MESSAGE).await;
+				}
+			}
+
+			// if the assignment is known to be valid, reward the peer
+			if entry.knowledge.known_messages.contains(&fingerprint) {
+				modify_reputation(ctx, peer_id.clone(), BENEFIT_VALID_MESSAGE).await;
+				entry.known_by.entry(peer_id).or_default().known_messages.insert(fingerprint.clone());
```

Comment:
I now think that we shouldn't do `or_default` here. I'm worried that peers could engage in this sequence:

1. Send a view update that removes block B from their view.
1. Send a message from `B` that they incur `COST_UNEXPECTED_MESSAGE` for, but then they receive `BENEFIT_VALID_MESSAGE`. This is a net negative reputation change. However, after this line, the `or_default()` adds them to the `known_by`.
1. Send all other messages related to `B`. They no longer incur `COST_UNEXPECTED_MESSAGE` but do receive `BENEFIT_VALID_MESSAGE`. This is a positive reputation change that may easily outweigh the negative in the last step.
1. Return to step 1.

In this way, peers could spam us and gain positive reputation at the same time.

Reply:
It'd be nice to test that this attack has a net negative reputation result.

Reply:
Good point, this `known_by` insertion was actually not in the guide, but I thought without it it would be possible to do another net positive attack. Probably the guide was correct.

Reply:
I think the addition of the `known_by` for peers who are already in the `HashMap` is correct. Just not for those who aren't yet in the map.